meson.build

```diff
@@ -336,14 +336,21 @@ if get_option('ta-leak-report')
     features += 'ta-leak-report'
 endif
 
-libdl = cc.find_library('libdl', required: false)
-if libdl.found()
-    dependencies += libdl
+libdl_found = false
+if cc.has_function('dlopen', prefix : '#include <dlfcn.h>')
     features += 'libdl'
+    libdl_found = true
+else
+    libdl = cc.find_library('libdl', required: false)
+    if libdl.found()
```

Comment:
You don't actually need this if/else. The libdl library can always be checked unconditionally and you can then use `cc.has_function` and give it libdl in the `dependencies` argument (whether the library is found or not). That should cover everything.

And I guess just a couple of naming nits. I'd prefer if we used `libdl_dep` (when trying to find the library) and `libdl` (when checking for the function).

Reply:
Thanks for the feedback!
As far as I understand, we have three possible variants:
1) The dlopen() function and its friends are part of libc and does not require libdl. The system itself comes without this library.
2) The dlopen() function and others are part of libc and does not require libdl. The 'libdl' stub library is present, which does not necessary need to be used.
3) The dlopen() function and others are directly in libdl. It is not possible to use libc instead.


Since we first check if the code that checks for dlopen function is building without using libdl, it seems to me that having the 'if-else' block here is correct: that way we will not link to libdl in the second case, i.e. when it is not mandatory.

Or am I missing something? My point is that if we already found dlopen in libc we don't need to think about having libdl at all, do we?
In any case, I'm willing to redo everything according to your recommendations, if necessary. At the same time, I would like to understand where I am wrong in my reasoning.

Reply:
You're right about the 3 different cases, but meson is pretty smart so you can handle them all the same way. Something like this:
```
libdl_dep = cc.find_library('libdl', required: false)
libdl = cc.has_function('dlopen', dependencies: libdl_dep, prefix : '#include <dlfcn.h>')
if libdl
    dependencies += libdl_dep
    features += 'libdl'
endif
```
would cover everything. If libdl_dep isn't found, then it's just an empty object that does nothing. If you have `dlopen` as part of your libc, the `has_function` check will still succeed.

Reply:
In case 2, the function check would succeed with or without `libdl_dep`, but checking for the library means the library will be passed as an unneeded argument to the linker, which will then try to link to it.

This is not a problem for GNU libc >=2.34, which does not provide libdl.so (it does provide libdl.so.2 which you can't link to with `-ldl`, but only satisfies libraries built on an older libc). The compat library it installs is only a static library, and libdl.a simply has no objects and cannot contribute code to the linker.
```bash
$ cat /usr/lib/libdl.a
!<arch>
```

The musl libc does the same thing.

I don't know whether BSD does the same trick for compat libraries. Probably it does, I'd be surprised if they actually had a shared library for it. I don't have a test system for it though.

So chances are the only difference between the two methods for probing dlopen support, is "the linker command line is slightly longer but achieves the exact same binary output".

Reply:
I will test it on OpenBSD where libdl is missing and share the results with you. Perhaps after that it will be easier to understand what the patch should look like.
Thank you very much for your feedback!

Reply:
@Dudemanguy's version works correctly both on Linux and OpenBSD.